File: ts/src/origins.ts
/**
 * Origins — ADR-049 §A.7.2 + ADR-047 §1.1.
 *
 * Partner backends call the unauthenticated proxy endpoints fronting
 * RealmID's platform-token-gated routes (e.g. `/auth/login`,
 * `/platforms/{id}/identity-providers`). Because RealmID no longer
 * inspects `Origin` on those routes, origin enforcement moves into the
 * SDK: this module fetches the per-realm origin allowlist and validates
 * a candidate `Origin` against it.
 *
 * Cache: in-memory, keyed by realmId, 5-minute TTL. On a 401 from the
 * underlying list call we invalidate the cached platform token, retry
 * once, then surface the error to the caller.
 */
import type { HttpClient } from "./http.js";
import { RealmError } from "./errors.js";
import { paginate, readPage, type Paginated, type PageOpts } from "./pagination.js";
import type { PlatformTokenManager } from "./platform-token-manager.js";

export interface Origin {
  id: string;
  domain: string;
  entity_type: "realm" | "tenant";
  entity_id: string;
  verification_id?: string | null;
  created_at?: string;
  detached_at?: string | null;
  [k: string]: unknown;
}

export interface OriginListOpts extends PageOpts {
  realmId: string;
}

export interface OriginValidateOpts {
  realmId: string;
  origin: string;
}

interface CacheEntry {
  allowlist: Set<string>;
  fetchedAt: number;
}

const CACHE_TTL_MS = 5 * 60 * 1000;

/**
 * Lowercase, strip scheme/port. Mirrors `domainmapping.Normalize` on
 * the Go side (ADR-049 §1.4).
 */
export function normalizeOrigin(raw: string): string {
  if (!raw) return "";
  let s = raw.trim().toLowerCase();
  if (s.startsWith("https://")) s = s.slice("https://".length);
  else if (s.startsWith("http://")) s = s.slice("http://".length);
  // strip path
  const slash = s.indexOf("/");
  if (slash >= 0) s = s.slice(0, slash);
  // strip port
  const colon = s.indexOf(":");
  if (colon >= 0) s = s.slice(0, colon);
  return s;
}

export class OriginsClient {
  private readonly cache = new Map<string, CacheEntry>();

  constructor(
    private readonly http: HttpClient,
    private readonly platformTokens: PlatformTokenManager,
    private readonly now: () => number = () => Date.now(),
  ) {}

  /**
   * Paginated list of live origin mappings for the realm. Wraps
   * `GET /platforms/{realmId}/origins` (ADR-049 §A.7.2). Platform-token
   * auth is attached automatically.
   */
  list(opts: OriginListOpts): Paginated<Origin> {
    if (!opts.realmId) {
      throw new RealmError({ code: "bad_request", message: "origins.list: realmId required" });
    }
    const realmId = opts.realmId;
    return paginate<Origin>(async (po) => {
      const raw = await this.callList(realmId, po);
      return readPage<Origin>(raw);
    });
  }

  /**
   * Validate the supplied origin against the realm's allowlist. Reads
   * from cache; on cache miss or TTL expiry, refetches the full
   * allowlist. Returns true iff the normalized origin appears in any
   * live `domain_mappings` row for the realm (across both
   * `entity_type='realm'` and `entity_type='tenant'`).
   */
  async validate(opts: OriginValidateOpts): Promise<boolean> {
    if (!opts.realmId) {
      throw new RealmError({ code: "bad_request", message: "origins.validate: realmId required" });
    }
    const norm = normalizeOrigin(opts.origin);
    if (!norm) return false;
    const allowlist = await this.getAllowlist(opts.realmId);
    return allowlist.has(norm);
  }

  /** Drop the cached allowlist for one realm (or all when omitted). */
  invalidate(realmId?: string): void {
    if (realmId) this.cache.delete(realmId);
    else this.cache.clear();
  }

  private async getAllowlist(realmId: string): Promise<Set<string>> {
    const hit = this.cache.get(realmId);
    if (hit && this.now() - hit.fetchedAt < CACHE_TTL_MS) {
      return hit.allowlist;
    }
    const fresh = await this.fetchAllowlist(realmId);
    this.cache.set(realmId, { allowlist: fresh, fetchedAt: this.now() });
    return fresh;
  }

  private async fetchAllowlist(realmId: string): Promise<Set<string>> {
    const set = new Set<string>();
    let cursor: string | undefined;
    while (true) {
      const raw = await this.callList(realmId, cursor ? { cursor } : {});
      const page = readPage<Origin>(raw);
      for (const row of page.items) {
        if (row.detached_at) continue;
        const d = normalizeOrigin(row.domain);
        if (d) set.add(d);
      }
      if (!page.nextCursor) return set;
      cursor = page.nextCursor;
    }
  }

  /**
   * Underlying list HTTP call with refresh-on-401 semantics. On 401 we
   * force-mint a fresh platform token and retry once; if still 401,
   * the error propagates to the caller (matches ADR-047 §1.1).
   */
  private async callList(realmId: string, opts: PageOpts): Promise<unknown> {
    try {
      return await this.http.request<unknown>({
        method: "GET",
        path: `/platforms/${encodeURIComponent(realmId)}/origins`,
        query: { cursor: opts.cursor, limit: opts.limit },
      });
    } catch (e) {
      if (e instanceof RealmError && e.code === "unauthorized") {
        this.platformTokens.invalidate();
        return await this.http.request<unknown>({
          method: "GET",
          path: `/platforms/${encodeURIComponent(realmId)}/origins`,
          query: { cursor: opts.cursor, limit: opts.limit },
        });
      }
      throw e;
    }
  }
}

File: go/origins.go
package realmid

import (
	ctxpkg "context"
	"strings"
	"sync"
	"time"
)

// Origin is one row in the per-realm origin allowlist (ADR-049 §A.7.2).
// Every listed row is live: the issuer's GET /platforms/{id}/origins filters
// detached rows server-side (detached_at IS NULL) and never serializes a
// detached_at field, so a returned row means the bare hostname routes to the
// referenced realm or tenant entity — and, by extension, is permitted to make
// unauthenticated proxy calls to the partner backend that fronts
// platform-token-gated RealmID routes.
type Origin struct {
	ID             string  `json:"id"`
	Domain         string  `json:"domain"`
	EntityType     string  `json:"entity_type"`
	EntityID       string  `json:"entity_id"`
	VerificationID *string `json:"verification_id,omitempty"`
	// CreatedAt is unix seconds. The issuer serializes every list/get
	// created_at as a JSON number (toDomainDTO → CreatedAt.Unix()); this was
	// mistyped as string through go/v0.21.0, which made the strict decode of
	// the origin allowlist throw on the BFF login hot path. Fixed in v0.22.0.
	CreatedAt int64 `json:"created_at,omitempty"`
}

// ListOriginsOptions parameterises Origins.List.
type ListOriginsOptions struct {
	RealmID string
	Cursor  string
	Limit   int
}

// ValidateOriginOptions parameterises Origins.Validate.
type ValidateOriginOptions struct {
	RealmID string
	Origin  string
}

// OriginsClient implements ADR-047 §1.1 — partner-side origin allowlist
// enforcement. Concurrent-safe.
type OriginsClient struct {
	realm *Realm
	now   func() time.Time

	mu    sync.RWMutex
	cache map[string]*originsCacheEntry
}

type originsCacheEntry struct {
	allowlist map[string]struct{}
	fetchedAt time.Time
}

// originsCacheTTL is the per-realm cache TTL (ADR-047 §1.1).
const originsCacheTTL = 5 * time.Minute

func newOriginsClient(r *Realm) *OriginsClient {
	now := time.Now
	if r != nil && r.cfg.Clock != nil {
		now = r.cfg.Clock
	}
	return &OriginsClient{
		realm: r,
		now:   now,
		cache: map[string]*originsCacheEntry{},
	}
}

// NormalizeOrigin lowercases an origin and strips scheme + port + path
// to a bare hostname. Mirrors `domainmapping.Normalize` server-side.
func NormalizeOrigin(raw string) string {
	if raw == "" {
		return ""
	}
	s := strings.ToLower(strings.TrimSpace(raw))
	if strings.HasPrefix(s, "https://") {
		s = s[len("https://"):]
	} else if strings.HasPrefix(s, "http://") {
		s = s[len("http://"):]
	}
	if i := strings.IndexByte(s, '/'); i >= 0 {
		s = s[:i]
	}
	if i := strings.IndexByte(s, ':'); i >= 0 {
		s = s[:i]
	}
	return s
}

// List returns a paginated iterator over the realm's live origin
// mappings. Wraps GET /platforms/{realmId}/origins (ADR-049 §A.7.2).
func (c *OriginsClient) List(_ ctxpkg.Context, opts ListOriginsOptions) (*Paginated[Origin], error) {
	if opts.RealmID == "" {
		return nil, &RealmError{Code: ErrCodeBadRequest, Message: "origins.List: RealmID required"}
	}
	realmID := opts.RealmID
	return newPaginated[Origin](func(ctx ctxpkg.Context, po PageOpts) (*Page[Origin], error) {
		return c.fetchPage(ctx, realmID, po)
	}), nil
}

// Validate reports whether the inbound origin is on the realm's
// allowlist. The cached allowlist is refreshed on cache miss + TTL
// expiry. On a 401 from the underlying list call, the platform token
// is invalidated and the call is retried once.
func (c *OriginsClient) Validate(ctx ctxpkg.Context, opts ValidateOriginOptions) (bool, error) {
	if opts.RealmID == "" {
		return false, &RealmError{Code: ErrCodeBadRequest, Message: "origins.Validate: RealmID required"}
	}
	norm := NormalizeOrigin(opts.Origin)
	if norm == "" {
		return false, nil
	}
	allow, err := c.allowlistFor(ctx, opts.RealmID)
	if err != nil {
		return false, err
	}
	_, ok := allow[norm]
	return ok, nil
}

// Invalidate drops the cached allowlist for one realm (or all if "").
func (c *OriginsClient) Invalidate(realmID string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if realmID == "" {
		c.cache = map[string]*originsCacheEntry{}
		return
	}
	delete(c.cache, realmID)
}

func (c *OriginsClient) allowlistFor(ctx ctxpkg.Context, realmID string) (map[string]struct{}, error) {
	c.mu.RLock()
	hit, ok := c.cache[realmID]
	c.mu.RUnlock()
	if ok && c.now().Sub(hit.fetchedAt) < originsCacheTTL {
		return hit.allowlist, nil
	}
	fresh, err := c.fetchAllowlist(ctx, realmID)
	if err != nil {
		return nil, err
	}
	c.mu.Lock()
	c.cache[realmID] = &originsCacheEntry{allowlist: fresh, fetchedAt: c.now()}
	c.mu.Unlock()
	return fresh, nil
}

func (c *OriginsClient) fetchAllowlist(ctx ctxpkg.Context, realmID string) (map[string]struct{}, error) {
	out := map[string]struct{}{}
	cursor := ""
	for {
		page, err := c.fetchPage(ctx, realmID, PageOpts{Cursor: cursor})
		if err != nil {
			return nil, err
		}
		for _, row := range page.Items {
			d := NormalizeOrigin(row.Domain)
			if d != "" {
				out[d] = struct{}{}
			}
		}
		if page.NextCursor == "" {
			return out, nil
		}
		cursor = page.NextCursor
	}
}

func (c *OriginsClient) fetchPage(ctx ctxpkg.Context, realmID string, po PageOpts) (*Page[Origin], error) {
	page, err := c.doListCall(ctx, realmID, po)
	if err == nil {
		return page, nil
	}
	// On 401, force a fresh platform token mint and retry once.
	if re, ok := err.(*RealmError); ok && re.Code == ErrCodeUnauthorized {
		c.realm.platformToken.invalidate()
		return c.doListCall(ctx, realmID, po)
	}
	return nil, err
}

func (c *OriginsClient) doListCall(ctx ctxpkg.Context, realmID string, po PageOpts) (*Page[Origin], error) {
	tok, err := c.realm.platformToken.get(ctx)
	if err != nil {
		return nil, err
	}
	q := map[string]string{}
	if po.Cursor != "" {
		q["cursor"] = po.Cursor
	}
	if po.Limit > 0 {
		// best-effort string convert without strconv import explosion
		q["limit"] = itoa(po.Limit)
	}
	var env pageEnvelope[Origin]
	if err := c.realm.http.do(ctx, requestOptions{
		Method: "GET",
		Path:   "/platforms/" + realmID + "/origins",
		Query:  q,
		Bearer: tok,
	}, &env); err != nil {
		return nil, err
	}
	return &Page[Origin]{Items: env.Items, NextCursor: env.NextCursor, Total: env.Total}, nil
}

// OriginClaim is the response from OriginsClient.Claim. Realm-origin
// claims are DNS-TXT only — the html_file method is not exposed at
// the platform surface (see ADR-049 §"Method options").
type OriginClaim struct {
	Domain         string `json:"domain"`
	Status         string `json:"status"`
	Method         string `json:"method"` // always "dns_txt"
	DNSRecordName  string `json:"dns_record_name"`
	DNSRecordValue string `json:"dns_record_value"`
}

// Claim begins a realm-owned origin claim. The DV row is owner-scoped
// to the realm — any admin in this realm can complete the verify step
// and add the bind. Re-claiming returns the existing pending token
// (idempotent).
func (c *OriginsClient) Claim(ctx ctxpkg.Context, realmID, hostname string) (*OriginClaim, error) {
	if realmID == "" || hostname == "" {
		return nil, &RealmError{Code: ErrCodeBadRequest, Message: "Origins.Claim: realmID and hostname required"}
	}
	tok, err := c.realm.platformToken.get(ctx)
	if err != nil {
		return nil, err
	}
	var resp OriginClaim
	if err := c.realm.http.do(ctx, requestOptions{
		Method: "POST",
		Path:   "/platforms/" + realmID + "/origins/claim",
		Bearer: tok,
		Body:   map[string]string{"domain": hostname},
	}, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// Verify drives the DNS check on the realm's pending DV row. Does NOT
// bind — call Origins.Bind (POST /platforms/{id}/origins) afterwards
// so a single verified apex can serve as the trust anchor for
// trusted-by-parent subdomain origins.
func (c *OriginsClient) Verify(ctx ctxpkg.Context, realmID, hostname string) (*DomainVerifyResult, error) {
	if realmID == "" || hostname == "" {
		return nil, &RealmError{Code: ErrCodeBadRequest, Message: "Origins.Verify: realmID and hostname required"}
	}
	tok, err := c.realm.platformToken.get(ctx)
	if err != nil {
		return nil, err
	}
	var resp DomainVerifyResult
	if err := c.realm.http.do(ctx, requestOptions{
		Method: "POST",
		Path:   "/platforms/" + realmID + "/origins/verify",
		Bearer: tok,
		Body:   map[string]string{"domain": hostname},
	}, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

func itoa(n int) string {
	if n == 0 {
		return "0"
	}
	neg := n < 0
	if neg {
		n = -n
	}
	var buf [20]byte
	i := len(buf)
	for n > 0 {
		i--
		buf[i] = byte('0' + n%10)
		n /= 10
	}
	if neg {
		i--
		buf[i] = '-'
	}
	return string(buf[i:])
}
